pkg/controller/machine_bootstrap_token.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+This file was copied and modified from the kubernetes/kubernetes project
+https://github.com/kubernetes/kubernetes/release-1.8/pkg/controller/deployment/util/pod_util.go
+
+Modifications Copyright (c) 2017 SAP SE or an SAP affiliate company. All rights reserved.
+*/
+
+// Package controller is used to provide the core functionalities of machine-controller-manager
+package controller
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/hex"
+	"math/big"
+	"regexp"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/gardener/machine-controller-manager/pkg/driver"
+	"github.com/golang/glog"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	bootstraptokenapi "k8s.io/cluster-bootstrap/token/api"
+	bootstraptokenutil "k8s.io/cluster-bootstrap/token/util"
+)
+
+const (
+	dirKubelet         = "/var/lib/kubelet"
+	bootstrapTokenPath = dirKubelet + "/" + "bootstrap-token"
+	bootstraptemplate  = `
+{{- define "CLOUD_CONFIG_WRITE_FILE" -}}
+- encoding: b64
+  content: {{ index . "CONTENT" }}
+  path: {{ index . "PATH" }}
+  permissions: "0644"
+{{- end -}}
+
+{{- define "SCRIPT_WRITE_FILE" -}}
+mkdir -p '{{ index . "DIR" }}'
+cat << EOF | base64 -d > '{{ index . "PATH" }}'
+{{ index . "CONTENT" }}
+EOF
+chmod '0644' '{{ index . "PATH" }}'
+{{- end -}}`
+)
+
+func (c *controller) addBootstrapTokenToUserData(machineName string, driver driver.Driver) error {
+	userData := driver.GetUserData()
+	glog.V(4).Infof("Creating bootstrap token!")
+	bootstrapTokenSecret, err := c.computeBootstrapToken(machineName)
+	if err != nil {
+		return err
+	}
+
+	token := bootstraptokenutil.TokenFromIDAndSecret(
+		string(bootstrapTokenSecret.Data[bootstraptokenapi.BootstrapTokenIDKey]),
+		string(bootstrapTokenSecret.Data[bootstraptokenapi.BootstrapTokenSecretKey]),
+	)
+	glog.V(4).Infof("Add bootstrap token %q to user-data!", token)
+	userData, err = addBootsrapTokenToUserData(userData, token)
+	if err != nil {
+		return err
+	}
+	driver.SetUserData(userData)
+	return nil
+}
+
+func (c *controller) computeBootstrapToken(machineName string) (secret *v1.Secret, err error) {
```

Comment:
Maybe rename this method to somehow indicate that a bootstrap token secret is created in the cluster if one does not exist.

Reply:
GetBootstrapTokenOrCreateItIfNotExist()?